### ISSUE TITLE
IOException encountered on AWS Batch

### ISSUE THREAD

**Issue by yihming** [aws-samples/aws-genomics-workflows#177]
To Whom It May Concern,

I followed your guide on creating Cromwell-based genomics workflow environment on AWS Batch via CloudFormation. However, when running my test, I got an IOException error:

```
[2021-09-17 16:15:42,27] [info] AwsBatchAsyncBackendJobExecutionActor [a2a1f270cumulus.cluster:NA:1]: job id: b1ff4ea5-2f7b-4576-8a83-abacb0712914
[2021-09-17 16:15:42,28] [info] AwsBatchAsyncBackendJobExecutionActor [a2a1f270cumulus.cluster:NA:1]: Status change from - to Initializing
[2021-09-17 16:18:55,62] [info] AwsBatchAsyncBackendJobExecutionActor [a2a1f270cumulus.cluster:NA:1]: Status change from Initializing to Running
[2021-09-17 16:20:40,93] [info] AwsBatchAsyncBackendJobExecutionActor [a2a1f270cumulus.cluster:NA:1]: Having to fall back to AWS query for status
[2021-09-17 16:20:42,26] [info] AwsBatchAsyncBackendJobExecutionActor [a2a1f270cumulus.cluster:NA:1]: Status change from Running to Failed
[2021-09-17 16:20:42,96] [info] WorkflowManagerActor: Workflow a2a1f270-4634-4fd4-b322-cd825303b151 failed (during ExecutingWorkflowState): cromwell.engine.io.IoAttempts$EnhancedCromwellIoException: [Attempted 1 time(s)] - IOException: Could not read from s3://gred-cumulus-dev/cromwell-execution/cumulus/a2a1f270-4634-4fd4-b322-cd825303b151/call-cluster/cluster-rc.txt: s3://s3.amazonaws.com/gred-cumulus-dev/cromwell-execution/cumulus/a2a1f270-4634-4fd4-b322-cd825303b151/call-cluster/cluster-rc.txt
Caused by: java.io.IOException: Could not read from s3://gred-cumulus-dev/cromwell-execution/cumulus/a2a1f270-4634-4fd4-b322-cd825303b151/call-cluster/cluster-rc.txt: s3://s3.amazonaws.com/gred-cumulus-dev/cromwell-execution/cumulus/a2a1f270-4634-4fd4-b322-cd825303b151/call-cluster/cluster-rc.txt
	at cromwell.core.path.EvenBetterPathMethods$$anonfun$fileIoErrorPf$1.applyOrElse(EvenBetterPathMethods.scala:117)
	at cromwell.core.path.EvenBetterPathMethods$$anonfun$fileIoErrorPf$1.applyOrElse(EvenBetterPathMethods.scala:116)
	at map @ cromwell.engine.io.nio.NioFlow.handleSingleCommand(NioFlow.scala:74)
	at map @ cromwell.engine.io.nio.NioFlow.$anonfun$processCommand$5(NioFlow.scala:60)
	at flatMap @ cromwell.engine.io.nio.NioFlow.$anonfun$processCommand$1(NioFlow.scala:59)
Caused by: java.nio.file.NoSuchFileException: s3://s3.amazonaws.com/gred-cumulus-dev/cromwell-execution/cumulus/a2a1f270-4634-4fd4-b322-cd825303b151/call-cluster/cluster-rc.txt
	at org.lerch.s3fs.S3FileSystemProvider.newInputStream(S3FileSystemProvider.java:355)
	at java.base/java.nio.file.Files.newInputStream(Files.java:156)
	at better.files.File.newInputStream(File.scala:455)
	at cromwell.core.path.BetterFileMethods.newInputStream(BetterFileMethods.scala:249)
	at cromwell.core.path.BetterFileMethods.newInputStream$(BetterFileMethods.scala:248)
	at cromwell.filesystems.s3.S3Path.newInputStream(S3PathBuilder.scala:157)
	at cromwell.core.path.EvenBetterPathMethods.mediaInputStream(EvenBetterPathMethods.scala:97)
	at cromwell.core.path.EvenBetterPathMethods.mediaInputStream$(EvenBetterPathMethods.scala:94)
	at cromwell.filesystems.s3.S3Path.mediaInputStream(S3PathBuilder.scala:157)
	at cromwell.core.path.EvenBetterPathMethods.$anonfun$withBufferedStream$1(EvenBetterPathMethods.scala:135)
	at cromwell.util.TryWithResource$.$anonfun$tryWithResource$1(TryWithResource.scala:14)
	at scala.util.Try$.apply(Try.scala:213)
	at cromwell.util.TryWithResource$.tryWithResource(TryWithResource.scala:10)
	at cromwell.core.path.EvenBetterPathMethods.withBufferedStream(EvenBetterPathMethods.scala:135)
	at cromwell.core.path.EvenBetterPathMethods.withBufferedStream$(EvenBetterPathMethods.scala:134)
	at cromwell.filesystems.s3.S3Path.withBufferedStream(S3PathBuilder.scala:157)
	at cromwell.core.path.EvenBetterPathMethods.limitFileContent(EvenBetterPathMethods.scala:142)
	at cromwell.core.path.EvenBetterPathMethods.limitFileContent$(EvenBetterPathMethods.scala:142)
	at cromwell.filesystems.s3.S3Path.limitFileContent(S3PathBuilder.scala:157)
	at cromwell.engine.io.nio.NioFlow.$anonfun$readAsString$1(NioFlow.scala:110)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:104)
	at cats.effect.internals.IORunLoop$RestartCallback.signal(IORunLoop.scala:463)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:484)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:422)
	at cats.effect.internals.IOShift$Tick.run(IOShift.scala:36)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:41)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:49)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

[2021-09-17 16:20:47,25] [info] WorkflowManagerActor: Workflow actor for a2a1f270-4634-4fd4-b322-cd825303b151 completed with status 'Failed'. The workflow will be removed from the workflow store.
[2021-09-17 16:21:38,79] [info] SingleWorkflowRunnerActor workflow finished with status 'Failed'.
[2021-09-17 16:21:42,22] [info] Workflow polling stopped
[2021-09-17 16:21:42,24] [info] Shutting down WorkflowStoreActor - Timeout = 5 seconds
[2021-09-17 16:21:42,25] [info] Aborting all running workflows.
[2021-09-17 16:21:42,25] [info] 0 workflows released by cromid-ba483c8
[2021-09-17 16:21:42,26] [info] Shutting down WorkflowLogCopyRouter - Timeout = 5 seconds
[2021-09-17 16:21:42,26] [info] Shutting down JobExecutionTokenDispenser - Timeout = 5 seconds
[2021-09-17 16:21:42,27] [info] WorkflowLogCopyRouter stopped
[2021-09-17 16:21:42,27] [info] WorkflowStoreActor stopped
[2021-09-17 16:21:42,27] [info] JobExecutionTokenDispenser stopped
[2021-09-17 16:21:42,28] [info] Shutting down WorkflowManagerActor - Timeout = 3600 seconds
[2021-09-17 16:21:42,28] [info] WorkflowManagerActor: All workflows finished
[2021-09-17 16:21:42,28] [info] WorkflowManagerActor stopped
[2021-09-17 16:21:42,61] [info] Connection pools shut down
[2021-09-17 16:21:42,61] [info] Shutting down SubWorkflowStoreActor - Timeout = 1800 seconds
[2021-09-17 16:21:42,61] [info] Shutting down JobStoreActor - Timeout = 1800 seconds
[2021-09-17 16:21:42,61] [info] SubWorkflowStoreActor stopped
[2021-09-17 16:21:42,62] [info] Shutting down CallCacheWriteActor - Timeout = 1800 seconds
[2021-09-17 16:21:42,62] [info] Shutting down ServiceRegistryActor - Timeout = 1800 seconds
[2021-09-17 16:21:42,62] [info] CallCacheWriteActor Shutting down: 0 queued messages to process
[2021-09-17 16:21:42,62] [info] CallCacheWriteActor stopped
[2021-09-17 16:21:42,62] [info] Shutting down DockerHashActor - Timeout = 1800 seconds
[2021-09-17 16:21:42,62] [info] Shutting down IoProxy - Timeout = 1800 seconds
[2021-09-17 16:21:42,63] [info] JobStoreActor stopped
[2021-09-17 16:21:42,64] [info] Shutting down connection pool: curAllocated=0 idleQueues.size=0 waitQueue.size=0 maxWaitQueueLimit=256 closed=false
[2021-09-17 16:21:42,64] [info] Shutting down connection pool: curAllocated=0 idleQueues.size=0 waitQueue.size=0 maxWaitQueueLimit=256 closed=false
[2021-09-17 16:21:42,64] [info] KvWriteActor Shutting down: 0 queued messages to process
[2021-09-17 16:21:42,64] [info] WriteMetadataActor Shutting down: 0 queued messages to process
[2021-09-17 16:21:42,65] [info] IoProxy stopped
[2021-09-17 16:21:42,65] [info] ServiceRegistryActor stopped
[2021-09-17 16:21:42,65] [info] DockerHashActor stopped
[2021-09-17 16:21:42,65] [info] Shutting down connection pool: curAllocated=0 idleQueues.size=0 waitQueue.size=0 maxWaitQueueLimit=256 closed=false
[2021-09-17 16:21:42,66] [info] Shutting down connection pool: curAllocated=0 idleQueues.size=0 waitQueue.size=0 maxWaitQueueLimit=256 closed=false
[2021-09-17 16:21:42,69] [info] Database closed
[2021-09-17 16:21:42,69] [info] Stream materializer shut down
[2021-09-17 16:21:42,70] [info] WDL HTTP import resolver closed
Workflow a2a1f270-4634-4fd4-b322-cd825303b151 transitioned to state Failed
```

When searching online, I found [this issue](https://github.com/broadinstitute/cromwell/issues/4341) is related to mine. However, all the links there are invalid, so I still cannot figure out a solution.

Could you please give me a hand? Thanks!

Sincerely,
Yiming

**Comment by multimeric** [comment]
So the core error here is `java.nio.file.NoSuchFileException: s3://s3.amazonaws.com/gred-cumulus-dev/cromwell-execution/cumulus/a2a1f270-4634-4fd4-b322-cd825303b151/call-cluster/cluster-rc.txt`. It looks like Cromwell has a malformed URL here, because `s3://` URLs don't start with `s3.amazonaws`. Only the HTTP versions of amazon URLs do that. So you likely should have used `s3://gred-cumulus-dev/` as your bucket path somewhere in the stack config. Refer to this page for more info: https://docs.aws.amazon.com/AmazonS3/latest/userguide/access-bucket-intro.html.

**Comment by yihming** [comment]
Thanks for helping me with this issue!

I follow the tutorial on creating my own distribution, which is located at `s3://gred-cumulus-cf-templates` bucket. Then I created stacks via CloudFormation console page, and changed the default template location info to direct to my own (please see screen shot below). `gred-cumulus-dev` is the S3 bucket storing analysis result, which was automatically created via CloudFormation.

![Screen Shot 2021-09-19 at 21 49 46](https://user-images.githubusercontent.com/2455754/133951523-cf82c740-3178-4e9f-b958-d0c84bcbc834.png)

![Screen Shot 2021-09-19 at 21 49 26](https://user-images.githubusercontent.com/2455754/133951459-396fc0e2-3cab-4471-a445-8b88adb8de71.png)

Is there anywhere else in the template config files that I should change from `s3.amazonaws.com` to my own?

**Comment by yihming** [comment]
The only thing I modified in my own template distribution is that I added the ECS-optimized AMI provided by my company's cloud team into `src/templates/gwfcore/gwfcore-launch-template.template.yaml` (see the line starting with "ImageId" below):

```
Resources:
  EC2LaunchTemplate:
    ... ...

      LaunchTemplateData:
        ImageId: "ami-0e403224d21661ede"
        # Used in tandem with UserData to check if the instance is provisioned
        # correctly. It is important to terminate mis-provisioned instances before
        # jobs are placed on them
        ... ...
```

**Comment by multimeric** [comment]
Hmm in that case it maybe is actually a bug in the stack. In which case I'll have to leave this to @wleepang. But again I think his (and AWS's) focus has moved more to the Genomics CLI so you might have better luck with applying for that.

**Comment by yihming** [comment]
Gotcha! Hope we could adapt to Genomics CLI soon. Thank you again for your help!

**Comment by ajfriedman18** [comment]
Looking through...Couple questions.

1. Since you created your own AMI, can you confirm that all of the UserData ran correctly? You could quickly just add a small min to you CE and then SSM connect into the EC2 instance. I wonder if it's throwing an error somewhere in the runcmd and you don't have the things installed that you should.
2. Can you confirm which AMI is actually being used in your Batch CE? Because Batch gives the CE precedence over the LT there might be a second clash here. https://github.com/aws-samples/aws-genomics-workflows/blob/c95a5f5d3cab349b838f14e2e4315e78ed910f92/src/templates/gwfcore/gwfcore-batch.template.yaml#L131

In general though, I would start looking at the AWS Genomics CLI for Cromwell support and would ping @wleepang for potential beta access

**Comment by yihming** [comment]
Hello @ajfriedman18 ,

Thank you for looking into my case. 

For the Compute Environment, since my tests only used the "OnDemand" one, I'll just list its information below:

![Screen Shot 2021-09-20 at 00 07 12](https://user-images.githubusercontent.com/2455754/133956278-353b2e99-811e-4422-916c-56d9d314e1c8.png)
![Screen Shot 2021-09-20 at 00 08 10](https://user-images.githubusercontent.com/2455754/133956335-a9e51329-01c1-4f68-9f7f-910bd783046e.png)

In summary, it seems this CE did use the AMI (`ami-0e403224d21661ede`) I set in the Launch Template yaml file.

I'll try your first suggestion by SSM into an EC2 instance created from my CE, and get back to you with results.

**Comment by yihming** [comment]
BTW, could you please explain a little bit more on "add a small min to you CE and then SSM connect into the EC2 instance"? I'm new to AWS, and have difficulty understanding how to check EC2 configs in the runtime. Many thanks!

**Comment by yihming** [comment]
@ajfriedman18 Thank you so much for your guidance. I've sent you a follow-up email, and will update here once it's resolved.
